mdelapenya/probot-codeowners-labeller
# Add Elastic's codeowners custom labeller

## What does this PR do?
It modifies the existing probot bot to read a Github CODEOWNERS file, using file spec: https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

For each modified path in a PR, it will look for its owner.

The owner is calculated in this manner (the original code used a regex for it, but I decided to use pure nodeJS code for it 🤷‍♀):
1. we parse each line in the CODEOWNERS file
1. if the line is a markdown comment, we do not process it and continue
1. then we tokenize the line by whitespaces (any number of occurrences)
1. for each token:
 - if the token starts with @, it's a user or team, so we add it as an owner
 - if the token is an email, we added it as an owner
 - if the token is not null and not empty, it's a file pattern (actual code)
1. for each file pattern in a line, we assign the owners based on above calculation
1. we massage the path, removing trailing `/` and adding `**` suffix to calculate a match between the modified file and the pattern, suing globs.
1. we finally obtain the latest owner, as the file allows overriding default, and this defined by the order they appear in the file.

For the custom Elastic labels, we define a map of string to string, with label names for Github usernames (maybe teams). This map is queried for each modified file in a PR, using the `codeowners.js` file.

## Why is it important?
We want to read the existing CODEOWNERS file to add a label in an automated manner.

As a drawback, and as consequence of not adding a new layer for the mapping between labels and teams in the own repository (possible at .github/labels.yml or similar), we have to maintain the bot's sourcecode when new labels/teams appear.
The bright side is that we centralise this information for every team under Elastic organisation, which is cool!

## Author's test list
- [x] NodeJS Eslint passing (npm run lint)
- [x] NodeJS Tests passing (npm run test)
- [x] Local Probot adding labels on my Kibana fork https://github.com/mdelapenya/kibana/pull/13

## How to test this
First of all, you have to set the Probot infra needed to start a probot, and connect it to a Github fork you own (preferable `elastic/kibana` because it has a CODEOWNERS file). To achieve this, please follow instructions here: https://github.com/elastic/infra/blob/master/flavortown/botelastic/CONTRIBUTING.md#setting-up
Apart from the Github App authorised for a fork you own, you'll need to export the environment variables generated in the `.env` file from that project:

```shell
$ export WEBHOOK_SECRET=
$ export PRIVATE_KEY=
$ export WEBHOOK_PROXY_URL=
$ export APP_ID=
```

Once exported those variables in the terminal for this project, please run `npm run start`, and create a PR in the kibana fork, modifying files under one of the paths described in the codeowners. 

>At this moment just APM, Logs&Infra, and App-Architecture are supported.

You'll see the labels appearing on the PR and the probot app will log it.

## Screenshots
Tests passing:
<img width="1112" alt="Screenshot 2019-12-17 at 09 18 09" src="https://user-images.githubusercontent.com/951580/70977443-38595b80-20ae-11ea-948d-3940a070c21a.png">

Probot logs with payload:
<img width="1114" alt="Screenshot 2019-12-17 at 09 17 54" src="https://user-images.githubusercontent.com/951580/70977442-38595b80-20ae-11ea-9f72-c72d9f5f722e.png">

Github labels added by bot:
<img width="791" alt="Screenshot 2019-12-17 at 09 18 19" src="https://user-images.githubusercontent.com/951580/70977445-38f1f200-20ae-11ea-8e25-aeb39796a291.png">
